### PR TITLE
DateTime types implicit conversion from varchar.

### PR DESCRIPTION
**Pull request by fiedukow**
Additionaly TIME_WITH_TIME_ZONE from VARCHAR missing conversion operator was implemented.

Missing bits:
- Coersion when cast from VARCHAR is ambiguous.
- Coersion to match INSERT types.

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java [hunk 1]
```diff
@@ -374,9 +386,15 @@ private static boolean typeParametersMatch(TypeSignatureParameter actualParamete
     public static Optional<TypeSignature> getCommonSuperTypeSignature(TypeSignature firstType, TypeSignature secondType)
     {
         TypeSignaturePair typeSignaturePair = new TypeSignaturePair(firstType, secondType);
-        if (typeSignaturePair.containsAnyOf(StandardTypes.VARCHAR, StandardTypes.DECIMAL)) {
+        if (typeSignaturePair.containsAnyOf(StandardTypes.DECIMAL)) {
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java:389]: Please revert the changes to this method, and just add the needed types here

```
for (String varcharSubType : new String[] {RegexpType.NAME, LikePatternType.NAME, JsonPathType.NAME}) {
             if (typeSignaturePair.is(StandardTypes.VARCHAR, varcharSubType)) {                  if (typeSignaturePair.is(StandardTypes.VARCHAR, varcharSubType)) {
                 return Optional.of(typeSignaturePair.get(varcharSubType));                      return Optional.of(typeSignaturePair.get(varcharSubType));
             }                   }
         }               }
```

File: presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java [hunk 2]
```diff
@@ -212,6 +212,12 @@ public static boolean canCoerce(List<? extends Type> actualTypes, List<Type> exp
         return true;
     }
 
+    private static boolean isDateTimeType(String type)
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java:215]: Maybe name it like `getVarcharSubtypes`, and make it return the list of types including `JsonPath` and etc. Than just use that list in both places `canCastTypeBase` and `getCommonSuperTypeForTypesWithLongParameters`

**Comment by fiedukow** [comment]
Continued in #129
